### PR TITLE
DAOS-14023 test: update md space tests for larger rdb

### PR DESCRIPTION
**Pull request by kccain**
With this change, test_metadata_fillup and other metadata space tests in server/metadata.py are updated to reflect the new default pool rdb capacity of 1GiB, rather than 128MiB. Also a dummy metadata workload function is added to spend time (150 seconds, pool property svc_ops_entry_age) warming up / populating rdb's svc_ops KVS. This is needed because the KVS (part of metadata duplicate operation detection feature) takes some rdb space and affects the amount of user-visible metadata resources seen. The tests are updated to exercise rdb storage limits under those more realistic conditions.

The configuration and test changes above result in longer execution times. So timeouts have been adjusted in metadata.yaml. Also, to conserve time, the test_metadata_fillup test is updated to avoid deleting each container after a pool's rdb has reached out of space. Finally, the test_metadata_addremove test is changed to perform fewer outer loops (4 instead of 10) to keep execution time reasonable.

The test_container_removal_after_der_nospace test is removed since it already overlapped with test_metadata_fillup and addremove testing. And because overall test times should be kept as low as possible.

Test-tag: test_metadata_fillup test_metadata_addremove test_metadata_server_restart
Test-repeat: 3
Skip-unit-tests: true
Skip-fault-injection-test: true

### Before requesting gatekeeper:

* [x] Two review approvals and any prior change requests have been resolved.
* [x] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [x] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [x] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Comment by kccain** [comment]
The initial commit of this PR contains the product code changes from a separate PR https://github.com/daos-stack/daos/pull/13866 that is intended to land first. So this will become a functional test only change after that.

**Review by dinghwah — COMMENTED** [review]
need to remove the unused import line 7

**Comment by kccain** [comment]
All tests passed. Also, here is a sampling of the additional time taken by the tests run here, compared to master branch weekly test build [369](https://build.hpdd.intel.com/job/daos-stack/job/daos/job/weekly-testing/369/artifact/Functional%20Hardware%20Large/server/metadata.py/results.json):

test_metadata_fillup: 351 seconds in PR [build 4](https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-13868/4/artifact/Functional%20Hardware%20Large/server/metadata.py/results.json), increase compared to 138 seconds in weekly [build 369](https://build.hpdd.intel.com/job/daos-stack/job/daos/job/weekly-testing/369/artifact/Functional%20Hardware%20Large/server/metadata.py/results.json).

test_metadata_addremove: 898 seconds in PR build 4, increase compared to 329 seconds in weekly build 369

test_metadata_server_restart: 445 seconds in PR build 4, about the same as 431 seconds in weekly build 369

PR seems ready for review so I will move it out of draft state.

**Review by liw — APPROVED** [review]

File: src/tests/ftest/server/metadata.py [hunk 1]
```diff
@@ -252,29 +289,43 @@ def test_metadata_fillup(self):
...
+                          self.CREATED_CONTAINERS_MIN)
+            self.fail("Phase 1: Created too few containers")
         self.log.info(
-            "Phase 3: passed (created %d / %d containers)", len(self.container), loop)
+            "Phase 1: passed (created %d / %d containers)", len(self.container), loop)
+
+        # Phase 2 clean up some containers (expected to succeed)
+        self.log.info("Phase 2: Cleaning up 500 containers (expected to work)")
+        if not self.destroy_n_containers(500):
+            self.fail("Phase 2: fail (unexpected container destroy error)")
```
Comment (daltonbohning) [inline src/tests/ftest/server/metadata.py:327]: Does this only destroy 500 to speed up the test?

File: src/tests/ftest/server/metadata.yaml [hunk 2]
```diff
@@ -15,41 +16,42 @@ server_config:
       nr_xs_helpers: 4
       first_core: 0
       pinned_numa_node: 0
-      fabric_iface: ib0
+      #fabric_iface: ib0
```
Comment (daltonbohning) [inline src/tests/ftest/server/metadata.yaml:19]: Debugging? Why is this removed?

File: src/tests/ftest/server/metadata.yaml [hunk 3]
```diff
@@ -15,41 +16,42 @@ server_config:
...
       env_vars:
-        - DAOS_MD_CAP=128
         - RDB_COMPACT_THRESHOLD=64
         - DD_MASK=group_metadata_only
       storage:
         0:
-          class: dcpm
-          scm_list: ["/dev/pmem0"]
+          #class: dcpm
+          #scm_list: ["/dev/pmem0"]
```
Comment (daltonbohning) [inline src/tests/ftest/server/metadata.yaml:29]: Debugging?

File: src/tests/ftest/server/metadata.yaml [hunk 4]
```diff
@@ -15,41 +16,42 @@ server_config:
...
+          class: ram
           scm_mount: /mnt/daos0
     1:
       targets: 8
       nr_xs_helpers: 4
       first_core: 0
       pinned_numa_node: 1
-      fabric_iface: ib1
+      #fabric_iface: ib1
```
Comment (daltonbohning) [inline src/tests/ftest/server/metadata.yaml:37]: Debugging?

File: src/tests/ftest/server/metadata.yaml [hunk 5]
```diff
@@ -15,41 +16,42 @@ server_config:
...
       env_vars:
-        - DAOS_MD_CAP=128
         - RDB_COMPACT_THRESHOLD=64
         - DD_MASK=group_metadata_only
       storage:
         0:
-          class: dcpm
-          scm_list: ["/dev/pmem1"]
+          #class: dcpm
+          #scm_list: ["/dev/pmem1"]
```
Comment (daltonbohning) [inline src/tests/ftest/server/metadata.yaml:47]: Debugging?

**Review by daltonbohning — CHANGES_REQUESTED** [review]
-1 for the debugging in `src/tests/ftest/server/metadata.yaml`

File: src/tests/ftest/server/metadata.py [hunk 6]
```diff
@@ -62,10 +61,14 @@ class ObjectMetadata(TestWithServers):
     """
 
     # Minimum number of containers that should be able to be created
-    CREATED_CONTAINERS_MIN = 2900
+    CREATED_CONTAINERS_MIN = 25000
```
Comment (dinghwah) [inline src/tests/ftest/server/metadata.py:64]: Could we add container to be created to metadata.yaml?

File: src/tests/ftest/server/metadata.py [hunk 7]
```diff
@@ -199,48 +256,28 @@ def test_metadata_fillup(self):
...
-        # periods of available space will occur, allowing a few container
-        # creates to succeed in the interim.
+        # Phase 1 sustained container creates even after no space error
         self.log.info(
-            "Phase 3: sustained container creates: to no space and beyond")
+            "Phase 1: sustained container creates: to no space and beyond")
         self.container = []
         sequential_fail_counter = 0
-        sequential_fail_max = 1000
+        sequential_fail_max = 512
```
Comment (dinghwah) [inline src/tests/ftest/server/metadata.py:278]: Please add sequential_fail_max var to metadata.yaml.

File: src/tests/ftest/server/metadata.py [hunk 8]
```diff
@@ -296,25 +347,28 @@ def test_metadata_addremove(self):
...
+
         self.container = []
         mean_cont_cnt = 0
         percent_cont = self.params.get("mean_percent", "/run/metadata/*")
 
         test_failed = False
         containers_created = []
-        for loop in range(10):
-            self.log.info("Container Create Iteration %d / 9", loop)
+        for loop in range(4):
```
Comment (dinghwah) [inline src/tests/ftest/server/metadata.py:359]: Please update test loop var to metadata.yaml.

File: src/tests/ftest/server/metadata.py [hunk 9]
```diff
@@ -62,10 +61,14 @@ class ObjectMetadata(TestWithServers):
...
 
     # Minimum number of containers that should be able to be created
-    CREATED_CONTAINERS_MIN = 2900
+    CREATED_CONTAINERS_MIN = 25000
+    # would be lower with engine env_vars DAOS_MD_CAP=128
+    # CREATED_CONTAINERS_MIN = 6500
 
     # Number of created containers that should not be possible
-    CREATED_CONTAINERS_LIMIT = 7500
+    CREATED_CONTAINERS_LIMIT = 36000
```
Comment (dinghwah) [inline src/tests/ftest/server/metadata.py:69]: Same as above. Please add created container limit var to metadata.yaml.

**Review by dinghwah — CHANGES_REQUESTED** [review]

File: src/tests/ftest/server/metadata.py (hunk 1 above)
Comment (kccain) [reply src/tests/ftest/server/metadata.py:327]: Yes indeed. It was a realization I made while considering the impact of changing the test to use the default rdb size, and recognizing that the point of the destroys is to make sure that destroys can occur when rdb is full. So the 500 containers approach here ensures what the test intended, while also trying to conserve execution time.

File: src/tests/ftest/server/metadata.yaml (hunk 2 above)
Comment (kccain) [reply src/tests/ftest/server/metadata.yaml:19]: oops, looks like I inadvertently pushed a version of this file that I use in a tmpfs only and non-IB development setup. Will fix this up in another push.

File: src/tests/ftest/server/metadata.py (hunk 6 above)
Comment (kccain) [reply src/tests/ftest/server/metadata.py:64]: I can look into this, may need some assistance if I can't figure it out based on looking at other similar examples in the code.

Comment (daltonbohning) [reply src/tests/ftest/server/metadata.py:64]: It works similar to this config:
https://github.com/daos-stack/daos/blob/042ffb9c9776757e59ec5424ad95d13eb34e600f/src/tests/ftest/server/metadata.yaml#L65-L66
And to get it in the test code:
https://github.com/daos-stack/daos/blob/042ffb9c9776757e59ec5424ad95d13eb34e600f/src/tests/ftest/server/metadata.py#L355
And that same location in the config is where it should go

File: src/tests/ftest/server/metadata.py [hunk 10]
```diff
@@ -184,6 +187,60 @@ def destroy_all_containers(self):
         self.container = []
         return len(errors) == 0
 
+    def destroy_n_containers(self, n):
```
[the remark this answers is not in the document]
Comment (kccain) [reply src/tests/ftest/server/metadata.py]: Changed n to num in the latest code on the branch.

File: src/tests/ftest/server/metadata.yaml (hunk 2 above)
Comment (kccain) [reply src/tests/ftest/server/metadata.yaml]: Fixed. Now just the intended diffs to this file are found in the latest code on the branch.

File: src/tests/ftest/server/metadata.py (hunk 6 above)
Comment (kccain) [reply src/tests/ftest/server/metadata.py]: added both parameters into that section of metadata.yaml in the latest code on the branch.

File: src/tests/ftest/server/metadata.py (hunk 7 above)
Comment (kccain) [reply src/tests/ftest/server/metadata.py]: added this in the latest code on the branch.

File: src/tests/ftest/server/metadata.py (hunk 8 above)
Comment (kccain) [reply src/tests/ftest/server/metadata.py]: added this in the latest code on the branch.

File: src/tests/ftest/server/metadata.py (hunk 9 above)
Comment (kccain) [reply src/tests/ftest/server/metadata.py]: added this in the latest code on the branch.

**Comment by kccain** [comment]
The PR build 6 single test [failure ](https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-13868/6/artifact/Functional%20Hardware%20Medium%20Verbs%20Provider/daos_test/suite.py/results.json)is unrelated and is documented in existing ticket https://daosio.atlassian.net/browse/DAOS-15124

File: src/tests/ftest/server/metadata.py (hunk 1 above)
Comment (daltonbohning) [reply src/tests/ftest/server/metadata.py]: Good decision! I think it's good to think of cases like this

File: src/tests/ftest/server/metadata.py [hunk 11]
```diff
@@ -184,6 +183,60 @@ def destroy_all_containers(self):
...
+
+        Returns:
+            bool: True if all of the containers were destroyed successfully;
+                False otherwise
+
+        """
+        if num > len(self.container):
+            num = len(self.container)
+        self.log.info("Destroying %d containers", num)
+        errors = self.destroy_containers(self.container[0:(num - 1)])
```
Comment (daltonbohning) [inline src/tests/ftest/server/metadata.py:200]: Sorry I didn't notice this before. This is actually off by one. The upper index is exclusive. For example
```
>>> my_list = [0,1,2,3]
>>> my_list[0:0]
[]
>>> my_list[0:1]
[0]
```
So you would want `[0:num]`

**Review by daltonbohning — APPROVED** [review]

**Review by liw — APPROVED** [review]

**Comment by kccain** [comment]
Testing completed, 3 repeat loops all passed. Ready for final code review.

https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-13868/12/artifact/Functional%20Hardware%20Large/server/metadata.py/repeat001/results.json
https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-13868/12/artifact/Functional%20Hardware%20Large/server/metadata.py/repeat002/results.json
https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-13868/12/artifact/Functional%20Hardware%20Large/server/metadata.py/repeat003/results.json

**Review by daltonbohning — APPROVED** [review]

**Review by dinghwah — APPROVED** [review]
